<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EBAME10: Functional enrichment in R • blog</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EBAME10: Functional enrichment in R">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ebame10_pt3.html">EBAME10: Differential Abundance with anvi'o ouput</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt1.html">EBAME10: Functional enrichment in anvi'o</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt2.html">EBAME10: Functional enrichment in R</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_extn_spaces.html">PNF: Extension spaces for phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_radEmu.html">PNF: radEmu for differential abundance</a></li>
    <li><a class="dropdown-item" href="../articles/stamps_2025.html">STAMPS 2025: How to write a great application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>EBAME10: Functional enrichment in R</h1>
                        <h4 data-toc-skip class="author">Amy D
Willis</h4>
            
            <h4 data-toc-skip class="date">October 17 2025</h4>
      

      <div class="d-none name"><code>ebame10_regression_pt2.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>First, work through the “EBAME10: Functional enrichment in
anvi’o”.</p>
<p>As discussed in lecture, the anvio functional enrichment toolkit is
great for quick and simple comparisons. However, should want more
details! How much do the odds of detecting a gene differ between the
groups? What about when you want to make more complex comparisons? What
about if you studied multiple sites longitudinally?</p>
<p>This tutorial walks you through pulling the data out of anvio, into
R, and running functional enrichment analyses yourself in R.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Open a terminal wherever you were running Part 1. You’re probably in
a directory called <code>Prochlorococcus_31_genomes</code>. I assume
anvio commands can still run.</p>
<p>Remember that we want to ask “How does the odds that a gene is
present differ between high and low light genomes?” To answer this, we
need to know which genes were detected in which genomes. So, let’s get
that info out of anvio and into a format we can load into R.</p>
<p>Let’s make a new directory to keep this info in.</p>
<pre><code>mkdir genes </code></pre>
<p>If you’ve setup your annotation databases correctly, you can pull out
the gene annotations and save them in the folder <code>genes</code> by
copying the following loop into your terminal (may take a few minutes,
so continue with setting R in the mean time):</p>
<pre><code>for file in *.db; do
    base=$(basename "$file" .db)
    anvi-export-functions -c "$file" -o "genes/${base}.txt"
done</code></pre>
<p>Ok! Time to open it in R. Open your favourite interactive R editor
and create a variable with the file path where the genomes DBs are (if
you already forgot, you can find it with <code>echo "$PWD"</code>). Load
in the high/low light data.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Change the following dir with what you obtain with echo above.</span></span>
<span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">"presentations/2510-ebame/Prochlorococcus_31_genomes/"</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"layer-additional-data.txt"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Note: If you get the message
<code>Error: '/.../...layer-additional-data.txt' does not exist.</code>,
double-check you replaced the directory address correctly.</p>
</div>
<div class="section level2">
<h2 id="wrangling-the-data-out-of-anvio-and-into-r">Wrangling the data out of anvi’o and into R<a class="anchor" aria-label="anchor" href="#wrangling-the-data-out-of-anvio-and-into-r"></a>
</h2>
<p>Our first step is to extract and format our data to something R
likes. We will do this in a few steps.</p>
<p>In the following code we:</p>
<ol style="list-style-type: decimal">
<li>List all files in the genes folder (extracted from anvi’o above)
with <code>list.files</code>, and for each file, we read as a table
(with <code>read_tsv</code>) and added a column named “sample” with the
name of the source file (with <code>mutate</code>).</li>
<li>Joined all these tables by their rows with
<code><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows()</a></code>.</li>
<li>Kept only those whose annotation source equals
<code>COG14_FUNCTION</code> with <code>filter</code>.</li>
<li>Remove the not-longer-needed columns “source” and “e-value” with
<code>select</code>.</li>
<li>Change the name of the column “function” with “cog_function” with
`rename``.</li>
<li>Made sure there are not identical rows using <code>distinct</code>
(repeated rows would have been deleted).</li>
</ol>
<pre><code><span><span class="va">what_is_there</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"genes"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"*.txt"</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span>, <span class="st">".txt"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">source</span> <span class="op">==</span> <span class="st">"COG14_FUNCTION"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">source</span>, <span class="op">-</span><span class="va">`e_value`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>cog_function <span class="op">=</span> <span class="va">`function`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">distinct</span></span></code></pre>
<p>“what_is_there” is now a table listing all genes found in each of the
sample genomes, along with their accession code and the functions
annotated to them by COG14_function.</p>
<p>Now, we will create a look-up table to connect accessions to
functions.</p>
<pre><code><span><span class="va">cog_fns</span> <span class="op">&lt;-</span> <span class="va">what_is_there</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">accession</span>, <span class="va">cog_function</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">distinct</span></span>
<span><span class="va">cog_fns</span></span></code></pre>
<p>How many distinct accession-function combinations we obtained?
<code>dim(cog_fns)</code></p>
<p>Now, we finish by constructing the table of gene presence/absence
with a useful format for the regression model we will fit later.</p>
<p>In the following code we construct our table “genes_long” from
“what_is_there” by:</p>
<ol style="list-style-type: decimal">
<li>Add a column called “presence” full of 1’s with <code>mutate</code>
(this eventually will indicate the gene in that row is present in the
sample on the same row)</li>
<li>Removing not-longer-useful columns and keeping only “sample”,
“accession” and “presence” with <code>select</code>
</li>
<li>Remove repeated rows with <code>distinct</code>
</li>
<li>Change the format of the table from long to wide, using “sample” as
the name of the rows, “accession” as the name of the columns and filling
up the table with “presence” with <code>pivot_longer</code>. Note that
<code>values_fill = 0</code> indicates that whenever a combination of
“sample”-“accession” was not a part of the table (ie we are missing the
value 1 in presence), to fill it up with 0. As a result, missing genes
in samples will have presence 0, while present genes will have presence
1.</li>
<li>Return to the long format with <code>pivot_longer</code>, keeping
all the zeros filled up in the previous step.</li>
<li>We add covariate information contained in the table “meta”, adding
the columns “clade” and “light” with <code>inner_join</code>.</li>
</ol>
<pre><code><span><span class="va">genes_long</span> <span class="op">&lt;-</span> <span class="va">what_is_there</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>presence <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample</span>,<span class="va">accession</span>,<span class="va">presence</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">distinct</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>values_from <span class="op">=</span> <span class="va">presence</span>, values_fill <span class="op">=</span> <span class="fl">0</span>, names_from <span class="op">=</span> <span class="va">accession</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">!</span><span class="va">sample</span>,names_to <span class="op">=</span> <span class="st">"cog"</span>, values_to <span class="op">=</span> <span class="st">"presence"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">meta</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span> <span class="op">=</span> <span class="st">"isolate"</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
<p>“genes_long” is now a table that contains a list of all the samples
and all the genes, with a 1 in “presence” if the gene is in the sample,
and 0 if it is missing, and with information on the sample by having
“clade” and “light”.</p>
</div>
<div class="section level2">
<h2 id="modelling-gene-presence">Modelling gene presence<a class="anchor" aria-label="anchor" href="#modelling-gene-presence"></a>
</h2>
<p>We are ready for some modeling!!</p>
<p>We will be employing the library raoBust, which implements
generalized linear models, along with robust score test. You may need to
install it if you haven’t.</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"statdivlab/raoBust"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">raoBust</span><span class="op">)</span></span></code></pre>
<p>Let’s start with a small case and turn our attention to the gene with
accession “COG4251”, because as shown by</p>
<pre><code><span><span class="va">cog_fns</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">accession</span> <span class="op">==</span> <span class="st">"COG4251"</span><span class="op">)</span> </span></code></pre>
<p>this gene codes for a light-sensitive protein. Thus, we would expect
the odds of it being present in a sample will be different in samples
collected with high or low light. How can we learn if this is the
case?</p>
<p>The following code fits a model to estimate the difference in odds we
are looking for.</p>
<pre><code><span><span class="va">genes_long</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG4251"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/raoBust/man/glm_test.html" class="external-link">glm_test</a></span><span class="op">(</span><span class="va">presence</span> <span class="op">~</span> <span class="va">light</span>, data <span class="op">=</span> <span class="va">.</span>,</span>
<span>           family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>What do we observe in the output? Is the estimate positive or
negative? Does it appear to be a significant signal (per the Robust
Score p-value)?</p>
<p>Take note of these things for now and contrast them with what we
obtain when exploring the presence/absence of a different gene. Let’s
explore the difference in the odd of presence of the gene with accession
“COG0592”.</p>
<pre><code><span><span class="va">genes_long</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/raoBust/man/glm_test.html" class="external-link">glm_test</a></span><span class="op">(</span><span class="va">presence</span> <span class="op">~</span> <span class="va">light</span>, data <span class="op">=</span> <span class="va">.</span>,</span>
<span>           family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>What happened here? Why could it be that the p-value obtained through
the Robust Score test is so large?</p>
<p>Well, it turns out…</p>
<pre><code><span><span class="va">genes_long</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">presence</span>,<span class="va">light</span><span class="op">)</span> </span></code></pre>
<p>Everyone has this gene! And that makes sense, because…</p>
<pre><code><span><span class="va">cog_fns</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">accession</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> </span></code></pre>
</div>
<div class="section level2">
<h2 id="interpreting-the-results">Interpreting the results<a class="anchor" aria-label="anchor" href="#interpreting-the-results"></a>
</h2>
<p>So we have found some evidence of relatedness between the amount of
light exposure in a sample and the presence/absence of gene “COG4251”,
while not having any evidence of the same thing for gene “COG0592”. How
can we properly report these finds?</p>
<p>Here is a paragraph that would make a statistician very happy to read
in your paper:</p>
<p>We estimate the odds of the gene “COG4251” being present in a
low-light ocean sample is 9.8^{-10} times lower than the odds of the
same gene being present in a high-light ocean sample (95% CI [2.9^{-10}
– 3.3^{-9}]; p = 3.32^{-03}). We have statistical evidence of a
difference in odds of the presence of this gene between samples from
low-light ocean areas and samples from high-light ocean areas.</p>
<p>The values reported above where obtained by computing the exponential
of the estimates and the confidence interval limits (for example,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>e</mi><mrow><mo>−</mo><mn>20.74839</mn></mrow></msup><mo>=</mo></mrow><annotation encoding="application/x-tex">e^{-20.74839} =</annotation></semantics></math>
9.8^{-10}), since the fitted model estimated the log fold-difference
between odds, and when reporting results it might be of interest to
express how the estimated odds directly change.</p>
<p>On the other hand, the reporting for gene “COG0592” would be
different. We are careful in reporting “negative” results, because the
lack of statistical evidence of a significant difference is not the same
as evidence for no difference at all.</p>
<p>We estimate the odds of the gene “COG0592” being present in a
low-light ocean sample is approximately the same as the odds of the same
gene being present in a high-light ocean sample (95% CI [0.47 – 2.1]; p
= 1.0). We have no statistical evidence to reject the hypothesis of the
odds of the presence of this gene being different between low-light
ocean samples and high-light ocean samples.</p>
</div>
<div class="section level2">
<h2 id="running-at-scale">Running at scale<a class="anchor" aria-label="anchor" href="#running-at-scale"></a>
</h2>
<p>How do we run and handle these tests for the whole set of genes?</p>
<p>For this we can create a table for each gene (using
<code>group_split</code> based in the column cog), and running the model
above for each. This may take a few minutes!</p>
<pre><code><span><span class="va">coefs</span> <span class="op">&lt;-</span> <span class="va">genes_long</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_split.html" class="external-link">group_split</a></span><span class="op">(</span><span class="va">cog</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span><span class="va">.</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">.x</span><span class="op">$</span><span class="va">cog</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/raoBust/man/glm_test.html" class="external-link">glm_test</a></span><span class="op">(</span><span class="va">presence</span> <span class="op">~</span> <span class="va">light</span>, data <span class="op">=</span> <span class="va">.x</span>, family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">imap_dfr</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">.x</span>, rownames <span class="op">=</span> <span class="st">"term"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>             <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>cog <span class="op">=</span> <span class="va">.y</span><span class="op">)</span>,</span>
<span>           .id <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="va">cog</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="exploring-and-interpreting-the-results">Exploring and interpreting the results<a class="anchor" aria-label="anchor" href="#exploring-and-interpreting-the-results"></a>
</h2>
<p>Now, look at the results! We can do this in several ways like:</p>
<ol style="list-style-type: decimal">
<li>Directly looking at the estimated numbers!</li>
<li>Plotting the estimated difference in odds, along with the robust
standard error.</li>
<li>We can order the estimated coefficients from those indicating higher
differences overall (some repetitions in our estimates happened. Why
could this be?)</li>
</ol>
<pre><code><span><span class="va">coefs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">`Non-robust Std Error`</span>, <span class="op">-</span><span class="va">`Non-robust Wald p`</span>, <span class="op">-</span><span class="va">`Robust Wald p`</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">coefs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Estimate</span>, y <span class="op">=</span> <span class="va">`Robust Std Error`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">coefs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>How can we make any conclusions on this large-scale exploration? And
how should we report these conclusions?</p>
<p>When fitting linear models and performing hypothesis tests at large
scale (as we are doing here, fitting a model for each of the 1546 gene
functions in our data), we need to be mindful on what we consider
“statistical significant”. Do this, or risk being overly confident of
noise!</p>
<p>Luckily, we have a good way to account for this volume of testing:
Q-values. These values control for the false discovery rate among all
tests, by describing the expected proportion of type 1 errors (rejecting
the null hypothesis when the null hypothesis is true) between the
rejected hypothesis.</p>
<p>In simple terms: if a test has q-value 0.05, it means we estimate 5%
of null hypothesis rejected for being as “extreme” as this hypothesis
would be false discoveries.</p>
<p>We can compute the q-values and add these to the estimated
coefficients for each test with the following code (you will need the
library qvalue):</p>
<pre><code><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"qvalue"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">coefs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">`Non-robust Std Error`</span>, <span class="op">-</span><span class="va">`Non-robust Wald p`</span>, <span class="op">-</span><span class="va">`Robust Wald p`</span>, <span class="op">-</span><span class="va">term</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># pull(`Robust Score p`) %&gt;%</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>qq <span class="op">=</span> <span class="fu">qvalue</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qvalue/man/qvalue.html" class="external-link">qvalue</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">`Robust Score p`</span><span class="op">)</span><span class="op">$</span><span class="va">qvalues</span><span class="op">)</span></span></code></pre>
<p>Doing this, what is the q-value associated to the estimator of the
difference in odds of the presence of gene with accession “COG4251”,
which we found significant in the previous section?</p>
<pre><code><span><span class="va">coefs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">`Non-robust Std Error`</span>, <span class="op">-</span><span class="va">`Non-robust Wald p`</span>, <span class="op">-</span><span class="va">`Robust Wald p`</span>, <span class="op">-</span><span class="va">term</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># pull(`Robust Score p`) %&gt;%</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>qq <span class="op">=</span> <span class="fu">qvalue</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qvalue/man/qvalue.html" class="external-link">qvalue</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">`Robust Score p`</span><span class="op">)</span><span class="op">$</span><span class="va">qvalues</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG4251"</span><span class="op">)</span></span></code></pre>
<p>We rejected the null hypothesis “The odds of the gene with accession
COG4251 being present in low-light ocean samples are the same as the
odds of the gene being present in samples from high-light ocean samples”
before. The q-value for this test (when ran along tests for all other
genes) is 0.0536. Meaning, only 5.36% of tests rejected under the same
conditions would be actually false discoveries. This is a positive
find!</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amy Willis, Sarah Teichman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
