<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EBAME10: Differential Abundance with anvi'o ouput • blog</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EBAME10: Differential Abundance with anvi'o ouput">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ebame10_pt3.html">EBAME10: Differential Abundance with anvi'o ouput</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt1.html">EBAME10: Functional enrichment in anvi'o</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt2.html">EBAME10: Functional enrichment in R</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_extn_spaces.html">PNF: Extension spaces for phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_radEmu.html">PNF: radEmu for differential abundance</a></li>
    <li><a class="dropdown-item" href="../articles/stamps_2025.html">STAMPS 2025: How to write a great application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>EBAME10: Differential Abundance with anvi'o ouput</h1>
                        <h4 data-toc-skip class="author">Maria Valdez,
Amy Willis</h4>
            
            <h4 data-toc-skip class="date">2025-10-17</h4>
      

      <div class="d-none name"><code>ebame10_pt3.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This lab covers “differential abundance” using metagenomics data,
emphasizing using the outputs of anvi’o’s
<code>anvi-estimate-scg-taxonomy</code> function. You could probably
adapt it to work with <code>anvi-profile-blitz</code>… let us know if
you want that!</p>
<p>It specifically covers the differential abundance method
<code>radEmu</code>. We like this method because it estimates changes in
the “absolute abundance” of microbial taxa, even without absolute
quantification data. Here, “absolute abundance” could be interpreted on
the cell count, cell concentration or DNA concentration scale. Yes! It’s
true! (We talked about this in lecture.)</p>
<p>We’ll illustrate its use on the the Healthy Dairy Workers dataset.
We’re interested in estimating changes in microbial abundances
<em>before</em> versus <em>after</em> commencing dairy work.</p>
<p>First let’s load libraries we’ll need. Install radEmu only if you
don’t already have it installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="co"># if (!("radEmu" %in% row.names(installed.packages()))) {</span></span>
<span><span class="co">#   remotes::install_github("statdivlab/radEmu")</span></span>
<span><span class="co"># }</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statdivlab/radEmu" class="external-link">radEmu</a></span><span class="op">)</span></span></code></pre></div>
<p>This is optional, but helpful to make some R packages be friends</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">select</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downloading-and-reading-in-metadata">Downloading and reading in metadata<a class="anchor" aria-label="anchor" href="#downloading-and-reading-in-metadata"></a>
</h2>
<p>You can download the data <a href="https://github.com/statdivlab/presentations/tree/main/ebame10/afternoon_data" class="external-link">here</a>.</p>
<p>The metadata is from SRA. Let’s read it in, and remove all columns
that are identical for all samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"../data/hdw_metadata.tsv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">as_tibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html" class="external-link">n_distinct</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   sample_id host_subject_id collection_date replicate.number</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> S200_V00  S200            2018-03-15                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> S200_V11  S200            2018-06-25                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> S200_V22  S200            2018-10-11                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> S200_V33  S200            2019-03-10                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> S200_V44  S200            2020-11-09                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> S201_V00  S201            2018-06-19                     1</span></span></code></pre>
<p>We can see the study collected samples at different points in time
(<code>collection_date</code>) from different hosts
(<code>host_subject_id</code>). How many hosts do we have? And how many
visits per host?</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">host_subject_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 9 × 2</span></span></span>
<span><span class="co">##   host_subject_id     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> S200                5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> S201                5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> S207                4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> S208                4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> S210                4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> S211                3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> S212                3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> S213                3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">9</span> S214                4</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="creating-and-loading-coverage-data">Creating and loading coverage data<a class="anchor" aria-label="anchor" href="#creating-and-loading-coverage-data"></a>
</h2>
<p>To save time, we’re not going to work with the sequences directly –
we did that for you in advance. But how did I make the data we’re going
to give you? I did coassemblies within subject, then ran
<code>anvi-run-scg-taxonomy</code> on each of my contigs databases. From
there, I estimated the taxonomy as follows:</p>
<pre><code>for file in 03_CONTIGS/*-contigs.db; do
    base=$(basename "$file" -contigs.db)
    profile="06_MERGED/${base}/PROFILE.db"
    output="11_MGX_TAXONOMY/${base}_S2.txt"
    anvi-estimate-scg-taxonomy -c $file -p $profile --compute-scg-coverages --metagenome-mode -o $output --scg-name-for-metagenome-mode Ribosomal_S2
done</code></pre>
<p>You can see that these results went in a folder called
<code>11_MGX_TAXONOMY</code>.</p>
<p>To run <code>radEmu</code>, we need a table with samples in the rows,
taxa in the columns, and entries corresponding to the coverages of each
taxon in each sample. So, we have to turn this data into that form.
We’re going to aggregate at the species level (i.e., we will sum the
coverage of different strains within the same species).</p>
<p>Here’s how you can do that (you can go through this slowly in your
own time, and adapt it to your own data). We remove some negative
control data; it’s not useful for us here.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coverages_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"../data/11_MGX_TAXONOMY/"</span>,pattern <span class="op">=</span> <span class="st">"^S.*txt"</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>FUN <span class="op">=</span> <span class="va">read_tsv</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">scg_name</span>, sep <span class="op">=</span> <span class="st">"_"</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_profile_db"</span>, <span class="cn">NA</span>, <span class="st">"marker"</span>, <span class="st">"strain"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">percent_identity</span>, <span class="op">-</span><span class="va">marker</span>, <span class="op">-</span><span class="va">strain</span>, <span class="op">-</span><span class="va">NEG_UWTrinh_BS_0001_F05</span>, <span class="op">-</span><span class="va">NEG_UWTrinh_BS_0001_H05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="va">S200_V00</span><span class="op">:</span><span class="va">S214_V2B</span>, </span>
<span>               names_to <span class="op">=</span> <span class="st">"sample_coverage"</span>, values_to <span class="op">=</span> <span class="st">"coverage"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_profile_db</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">sample_coverage</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_profile_db</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">sample_coverage</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">t_domain</span>, <span class="va">t_phylum</span>, <span class="va">t_class</span>, <span class="va">t_order</span>, <span class="va">t_family</span>, <span class="va">t_genus</span>, <span class="va">t_species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">coverage</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"d__"</span>, <span class="va">t_domain</span>, <span class="st">";p__"</span>, <span class="va">t_phylum</span>, </span>
<span>                       <span class="st">";c__"</span>, <span class="va">t_class</span>, <span class="st">";o__"</span>,<span class="va">t_order</span>,<span class="st">";f__"</span>,<span class="va">t_family</span>,</span>
<span>                       <span class="st">";g__"</span>, <span class="va">t_genus</span>, <span class="st">";s__"</span>,<span class="va">t_species</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sp_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">t_species</span> <span class="op">==</span> <span class="st">"None"</span>, <span class="va">taxa</span>, <span class="va">t_species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">coverages</span> <span class="op">&lt;-</span> <span class="va">coverages_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">coverage</span>, <span class="va">sp_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>values_from <span class="op">=</span> <span class="va">coverage</span>, values_fill <span class="op">=</span> <span class="fl">0</span>, names_from <span class="op">=</span> <span class="va">sp_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span> </span></code></pre></div>
<p>How many species are there?</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  35 472</span></span></code></pre>
<p>Based on the dimensions of the metadata and coverage data, how are
these data tables related?</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "S200_V00" "S200_V11" "S200_V22" "S200_V33" "S200_V44" "S201_V00"</span></span>
<span><span class="co">##  [7] "S201_V11" "S201_V22" "S201_V33" "S201_V44" "S207_V0A" "S207_V0B"</span></span>
<span><span class="co">## [13] "S207_V2A" "S207_V2B" "S208_V0A" "S208_V0B" "S208_V2A" "S208_V2B"</span></span>
<span><span class="co">## [19] "S210_V0A" "S210_V0B" "S210_V1A" "S210_V1B" "S211_V00" "S211_V11"</span></span>
<span><span class="co">## [25] "S211_V22" "S212_V00" "S212_V11" "S212_V22" "S213_V00" "S213_V11"</span></span>
<span><span class="co">## [31] "S213_V22" "S214_V0A" "S214_V0B" "S214_V2A" "S214_V2B"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span><span class="op">$</span><span class="va">sample_id</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "S200_V00" "S200_V11" "S200_V22" "S200_V33" "S200_V44" "S201_V00"</span></span>
<span><span class="co">##  [7] "S201_V11" "S201_V22" "S201_V33" "S201_V44" "S207_V0A" "S207_V0B"</span></span>
<span><span class="co">## [13] "S207_V2A" "S207_V2B" "S208_V0A" "S208_V0B" "S208_V2A" "S208_V2B"</span></span>
<span><span class="co">## [19] "S210_V0A" "S210_V0B" "S210_V1A" "S210_V1B" "S211_V00" "S211_V11"</span></span>
<span><span class="co">## [25] "S211_V22" "S212_V00" "S212_V11" "S212_V22" "S213_V00" "S213_V11"</span></span>
<span><span class="co">## [31] "S213_V22" "S214_V0A" "S214_V0B" "S214_V2A" "S214_V2B"</span></span></code></pre>
<p>To keep things simple, if there are technical replicates, we will
retain only one. For this, we will remove all samples labelled with a
“B” at the end.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coverages</span> <span class="op">&lt;-</span> <span class="va">coverages</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">endsWith</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/startsWith.html" class="external-link">endsWith</a></span><span class="op">(</span><span class="va">sample_id</span>,<span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Remember that we’re interested in estimating fold-changes in
microbial abundances before compared to after commencing dairy work. So,
we will create a variable called <code>exposed</code>, which will
differentiate between samples collected at baseline (FALSE: very first
sample collected for each host) and those collected after exposure
(TRUE: samples collected subsequently).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"Rownames"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">host_subject_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>exposed <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">collection_date</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">collection_date</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var<span class="op">=</span><span class="st">"Rownames"</span><span class="op">)</span></span></code></pre></div>
<p>Check we have one exposed and the rest unexposed:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">host_subject_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">exposed</span>,<span class="va">host_subject_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 18 × 3</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   host_subject_id [9]</span></span></span>
<span><span class="co">##    host_subject_id exposed     n</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> S200            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> S200            TRUE        4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> S201            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> S201            TRUE        4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> S207            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> S207            TRUE        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> S208            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> S208            TRUE        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> S210            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> S210            TRUE        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> S211            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> S211            TRUE        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span> S212            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">14</span> S212            TRUE        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">15</span> S213            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">16</span> S213            TRUE        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">17</span> S214            FALSE       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">18</span> S214            TRUE        1</span></span></code></pre>
<p>So that <code>radEmu</code> knows now to connect the metadata and the
coverage data, we need to put the sample labels in the row names:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">$</span><span class="va">sample_id</span></span></code></pre></div>
<p>Finally, to run radEmu, we need to confirm that all species are
detected in at least one sample. (Even if it was true at the beginning,
we could have taxa that were only in the removed samples.)</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coverages</span> <span class="op">&lt;-</span> <span class="va">coverages</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  27 469</span></span></code></pre>
<p>27 samples, 469 species, woohoo!</p>
<p>Phew! That was a lot of data processing! That’s usually most of the
work!</p>
<p>Now, we can estimate fold-differences.</p>
</div>
<div class="section level2">
<h2 id="fitting-a-model">Fitting a model<a class="anchor" aria-label="anchor" href="#fitting-a-model"></a>
</h2>
<p>The function that we use to fit our model is called
<code>emuFit</code>. Here’s the arguments that we will use:</p>
<ul>
<li>
<code>formula</code>: This is a formula telling <code>emuFit</code>
what predictors to use. We are using just <code>exposed</code>, but your
model could be much more complicated.</li>
<li>
<code>data</code>: The metadata table.</li>
<li>
<code>Y</code>: A matrix or data frame containing our observed
abundance data. For us, it’s the coverage table. The rows give the
observations (samples), and the columns give the categories
(species).</li>
<li>
<code>cluster</code>: (Optional) If samples are correlated, the name
of the ‘’groups’’ to which each sample belongs to, making them not
independent. In our case, samples from the same host form clusters.</li>
</ul>
<p>There’s one more important argument to know about:</p>
<ul>
<li>
<code>run_score_tests</code>: (Optional) A logical value denoting
whether or not to run score tests.</li>
</ul>
<p>Say NO for now! We’ll come back to this.</p>
<p>Let’s see how the model fitting looks with the variables explained
above! This code should take less than a minute to run.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span><span class="va">hdw_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">exposed</span>,</span>
<span>                  data <span class="op">=</span> <span class="va">meta</span>,</span>
<span>                  Y <span class="op">=</span> <span class="va">coverages</span>,</span>
<span>                  cluster <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">host_subject_id</span>,</span>
<span>                  run_score_tests <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##  11.209   0.294  12.056</span></span></code></pre>
<p>Let’s look at our results and talk through them together.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdw_fit</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## emuFit(Y = coverages, formula = ~exposed, data = meta, cluster = meta$host_subject_id, </span></span>
<span><span class="co">##     run_score_tests = FALSE)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficient estimates with the largest magnitudes:</span></span>
<span><span class="co">##       covariate</span></span>
<span><span class="co">## 4   exposedTRUE</span></span>
<span><span class="co">## 395 exposedTRUE</span></span>
<span><span class="co">## 55  exposedTRUE</span></span>
<span><span class="co">## 3   exposedTRUE</span></span>
<span><span class="co">## 101 exposedTRUE</span></span>
<span><span class="co">## 50  exposedTRUE</span></span>
<span><span class="co">## 353 exposedTRUE</span></span>
<span><span class="co">## 11  exposedTRUE</span></span>
<span><span class="co">## 273 exposedTRUE</span></span>
<span><span class="co">## 56  exposedTRUE</span></span>
<span><span class="co">## 433 exposedTRUE</span></span>
<span><span class="co">## 15  exposedTRUE</span></span>
<span><span class="co">## 95  exposedTRUE</span></span>
<span><span class="co">## 160 exposedTRUE</span></span>
<span><span class="co">## 77  exposedTRUE</span></span>
<span><span class="co">## 448 exposedTRUE</span></span>
<span><span class="co">## 232 exposedTRUE</span></span>
<span><span class="co">## 173 exposedTRUE</span></span>
<span><span class="co">## 290 exposedTRUE</span></span>
<span><span class="co">## 109 exposedTRUE</span></span>
<span><span class="co">##                                                                                                                  category</span></span>
<span><span class="co">## 4                                                                                                 Bifidobacterium_bifidum</span></span>
<span><span class="co">## 395                                                                                                 Dialister_sp002320515</span></span>
<span><span class="co">## 55                                                                                                  Ruminococcus_B_gnavus</span></span>
<span><span class="co">## 3                                                                                                Bifidobacterium_animalis</span></span>
<span><span class="co">## 101                                                                                              Phocaeicola_massiliensis</span></span>
<span><span class="co">## 50                     d__Bacteria;p__Bacillota_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Merdisoma;s__None</span></span>
<span><span class="co">## 353                                                                                                  CALXCQ01_sp944386835</span></span>
<span><span class="co">## 11          d__Bacteria;p__Actinomycetota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Eggerthella;s__None</span></span>
<span><span class="co">## 273                                                                                                   CAG-353_sp900066885</span></span>
<span><span class="co">## 56                   d__Bacteria;p__Bacillota_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Ventrimonas;s__None</span></span>
<span><span class="co">## 433                                                                                             Angelakisella_sp900552845</span></span>
<span><span class="co">## 15                d__Bacteria;p__Bacillota;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;g__Thomasclavelia;s__None</span></span>
<span><span class="co">## 95                                                                                                   Megasphaera_elsdenii</span></span>
<span><span class="co">## 160  d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__None</span></span>
<span><span class="co">## 77                                                                                                 Vescimonas_sp000435975</span></span>
<span><span class="co">## 448                                                                                                 Merdousia_sp900548275</span></span>
<span><span class="co">## 232                                                                                                  UMGS1901_sp900553755</span></span>
<span><span class="co">## 173                       d__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Weissella;s__None</span></span>
<span><span class="co">## 290 d__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Citrobacter;s__None</span></span>
<span><span class="co">## 109                                                                                                 Alistipes_onderdonkii</span></span>
<span><span class="co">##     category_num  estimate        se     lower     upper</span></span>
<span><span class="co">## 4              4  7.558651 0.5347523  6.510556  8.606746</span></span>
<span><span class="co">## 395          395  6.544570 0.9669288  4.649425  8.439716</span></span>
<span><span class="co">## 55            55  5.613179 0.9524010  3.746507  7.479850</span></span>
<span><span class="co">## 3              3 -5.463178 0.8907733 -7.209062 -3.717295</span></span>
<span><span class="co">## 101          101  5.215436 0.9513605  3.350804  7.080068</span></span>
<span><span class="co">## 50            50  5.196918 0.4440557  4.326585  6.067252</span></span>
<span><span class="co">## 353          353  5.032484 0.1982103  4.643999  5.420969</span></span>
<span><span class="co">## 11            11  5.015379 0.9506583  3.152123  6.878636</span></span>
<span><span class="co">## 273          273 -4.997699 0.9624874 -6.884140 -3.111259</span></span>
<span><span class="co">## 56            56  4.936816 0.9503418  3.074180  6.799451</span></span>
<span><span class="co">## 433          433 -4.873355 1.0007881 -6.834863 -2.911846</span></span>
<span><span class="co">## 15            15  4.820096 0.9498232  2.958476  6.681715</span></span>
<span><span class="co">## 95            95  4.657158 0.9489901  2.797172  6.517145</span></span>
<span><span class="co">## 160          160  4.580213 0.8837655  2.848065  6.312362</span></span>
<span><span class="co">## 77            77  4.545697 0.9483368  2.686991  6.404403</span></span>
<span><span class="co">## 448          448 -4.444996 0.9952633 -6.395676 -2.494315</span></span>
<span><span class="co">## 232          232 -4.347533 0.4432254 -5.216239 -3.478827</span></span>
<span><span class="co">## 173          173 -4.329899 0.9239342 -6.140777 -2.519021</span></span>
<span><span class="co">## 290          290 -4.307836 0.9534411 -6.176546 -2.439126</span></span>
<span><span class="co">## 109          109  4.290305 0.2800291  3.741458  4.839152</span></span></code></pre>
<pre><code><span><span class="co">## To obtain the entire coefficient table, use the command `emuFit_object$coef`.</span></span></code></pre>
<p>The printed output is sorted by the largest estimated effect sizes
(fold differences), but how do we look at a specific species? The way to
access estimated coefficients and confidence intervals from the model is
with <code>hdw_fit$coef</code>. Let’s look at one:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">52</span>, <span class="op">]</span> </span></code></pre></div>
<pre><code><span><span class="co">##      covariate                   category category_num estimate         se</span></span>
<span><span class="co">## 52 exposedTRUE Oliverpabstia_intestinalis           52 2.411388 0.06977598</span></span>
<span><span class="co">##      lower    upper score_stat pval</span></span>
<span><span class="co">## 52 2.27463 2.548147         NA   NA</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="interpreting-the-results">Interpreting the results<a class="anchor" aria-label="anchor" href="#interpreting-the-results"></a>
</h2>
<p>Let’s interpret these results! Here’s three equivalent
interpretations:</p>
<ul>
<li>We estimate that the average abundance of Oliverpabstia_intestinalis
in metagenomes is 11 times greater after commencing dairy work, when
compared to the typical fold-differences in the average abundance of
taxa across these visits. (Yep – that’s a <em>ratio</em> of
<em>ratios</em>.)</li>
</ul>
<!-- - We estimate that the abundance of *Oliverpabstia intestinalis* in metagenomes from subsequent visits is $e^{2.41} = 11.13$ times greater than those from the baseline, when compared to the typical fold-differences in the abundance of taxa across these visits. (Yep -- that's a *ratio* of *ratios*.) --><ul>
<li><p>We estimate that the log-fold difference in the average abundance
of Oliverpabstia_intestinalis in metagenomes from visits post and prior
exposure is 2.4 greater than the typical log-fold difference in the
average abundance of taxa across these groups. (That’s a
<em>difference</em> in <em>differences</em>.)</p></li>
<li><p>Under the assumption that most taxa do not differ in average
abundance between visits post and prior exposure, we estimate that the
abundance of Oliverpabstia_intestinalis in metagenomes from post
exposure visits is 11 times greater than prior exposure visits.</p></li>
</ul>
<p>By default <code>radEmu</code> compares the log fold difference in
one taxon to the <em>typical</em> (approximately the median) log fold
difference in all taxa. However, <code>radEmu</code> can estimate other
types of parameters. For example, we can compare log fold differences to
the log fold difference for a specific (or reference) taxon. <a href="https://statdivlab.github.io/radEmu/articles/radEmu_with_reference_taxon.html" class="external-link">This
vignette</a> shows you how to modify your code to use a reference
taxon.</p>
<p><strong>Please take the time to interpret your estimates, don’t just
report p-values.</strong></p>
<p>Ok, that’s one taxon. Let’s look at our results.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>order <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">order</span>, y <span class="op">=</span> <span class="va">estimate</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">lower</span>, ymax <span class="op">=</span> <span class="va">upper</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">"Estimate (with 95% confidence interval)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Log fold-difference estimates"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ebame10_pt3_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Here we can see the distribution of log fold-difference estimates
from our model, as well as 95% confidence intervals.</p>
<div class="section level3">
<h3 id="is-this-even-reasonable">Is this even reasonable?<a class="anchor" aria-label="anchor" href="#is-this-even-reasonable"></a>
</h3>
<p>A quick check…</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7.558651</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1917.257</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">largest_diff</span> <span class="op">&lt;-</span> <span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">tibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">category</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/add_column.html" class="external-link">add_column</a></span><span class="op">(</span>coverage <span class="op">=</span> <span class="va">coverages</span><span class="op">[[</span><span class="va">largest_diff</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">exposed</span>, <span class="va">coverage</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          sample_id exposed  coverage</span></span>
<span><span class="co">## S200_V00  S200_V00   FALSE   0.00000</span></span>
<span><span class="co">## S200_V11  S200_V11    TRUE  66.49854</span></span>
<span><span class="co">## S200_V22  S200_V22    TRUE 252.01276</span></span>
<span><span class="co">## S200_V33  S200_V33    TRUE 244.21329</span></span>
<span><span class="co">## S200_V44  S200_V44    TRUE   0.00000</span></span>
<span><span class="co">## S201_V00  S201_V00   FALSE   0.00000</span></span>
<span><span class="co">## S201_V11  S201_V11    TRUE   0.00000</span></span>
<span><span class="co">## S201_V22  S201_V22    TRUE   0.00000</span></span>
<span><span class="co">## S201_V33  S201_V33    TRUE   0.00000</span></span>
<span><span class="co">## S201_V44  S201_V44    TRUE   0.00000</span></span>
<span><span class="co">## S207_V0A  S207_V0A   FALSE   0.00000</span></span>
<span><span class="co">## S207_V2A  S207_V2A    TRUE   0.00000</span></span>
<span><span class="co">## S208_V0A  S208_V0A   FALSE   0.00000</span></span>
<span><span class="co">## S208_V2A  S208_V2A    TRUE   0.00000</span></span>
<span><span class="co">## S210_V0A  S210_V0A   FALSE   0.00000</span></span>
<span><span class="co">## S210_V1A  S210_V1A    TRUE   0.00000</span></span>
<span><span class="co">## S211_V00  S211_V00   FALSE   0.00000</span></span>
<span><span class="co">## S211_V11  S211_V11    TRUE 363.30792</span></span>
<span><span class="co">## S211_V22  S211_V22    TRUE 192.85163</span></span>
<span><span class="co">## S212_V00  S212_V00   FALSE   0.00000</span></span>
<span><span class="co">## S212_V11  S212_V11    TRUE   0.00000</span></span>
<span><span class="co">## S212_V22  S212_V22    TRUE 259.47017</span></span>
<span><span class="co">## S213_V00  S213_V00   FALSE   0.00000</span></span>
<span><span class="co">## S213_V11  S213_V11    TRUE   0.00000</span></span>
<span><span class="co">## S213_V22  S213_V22    TRUE   0.00000</span></span>
<span><span class="co">## S214_V0A  S214_V0A   FALSE   0.00000</span></span>
<span><span class="co">## S214_V2A  S214_V2A    TRUE   0.00000</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="hypothesis-tests-and-p-values">Hypothesis tests and p-values<a class="anchor" aria-label="anchor" href="#hypothesis-tests-and-p-values"></a>
</h2>
<p>Now, let’s talk about testing the null hypothesis that a species
doesn’t differ in its average abundance before vs after exposure. While
its tempting to use the confidence intervals to just say “p&lt;0.05” or
“not p&lt;0.05”, there’s a MUCH more reliable way. Robust score tests
are the reliable option – but they also take some time.</p>
<p>Let’s start by getting one p-value, then get them all.</p>
<p>To set up this test, we can again run <code>emuFit</code>, giving it
the fitted values that it has already found:</p>
<ul>
<li>
<code>formula</code>, <code>data</code> and <code>Y</code> are as
before</li>
<li>
<code>hdw_fit</code> is our previous fitted object (the output of
<code>emuFit</code>)</li>
<li>
<code>test_kj</code> a data frame listing the indices of the
parameters (in <code>hdw_fit$B</code>) that we want to test.</li>
</ul>
<p>This should take about a minute to run on the RStudio server.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa_to_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">134</span>, <span class="fl">372</span><span class="op">)</span></span>
<span><span class="va">covariate_to_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="st">"exposedTRUE"</span> <span class="op">==</span> <span class="va">hdw_fit</span><span class="op">$</span><span class="va">B</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">rownames</span><span class="op">)</span></span>
<span><span class="va">two_robust_score_tests</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">exposed</span>,</span>
<span>                                 data <span class="op">=</span> <span class="va">meta</span>,</span>
<span>                                 Y <span class="op">=</span> <span class="va">coverages</span>,</span>
<span>                                 cluster <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">host_subject_id</span>,</span>
<span>                                 <span class="co"># we already have hdw_fit, don't recompute this </span></span>
<span>                                 fitted_model <span class="op">=</span> <span class="va">hdw_fit</span>,</span>
<span>                                 refit <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                 test_kj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>k <span class="op">=</span> <span class="va">covariate_to_test</span>,</span>
<span>                                                      j <span class="op">=</span> <span class="va">taxa_to_test</span><span class="op">)</span>,</span>
<span>                                 <span class="co"># have it tell you when it runs score tests</span></span>
<span>                                 verbose <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                 <span class="co"># don't need to re-compute confidence intervals</span></span>
<span>                                 compute_cis <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<p>Let’s take a look at the test output.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">two_robust_score_tests</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="va">taxa_to_test</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"pval"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2143957</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="va">taxa_to_test</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       covariate                category category_num  estimate        se</span></span>
<span><span class="co">## 134 exposedTRUE Clostridium_sp900543325          134 -2.158088 1.0302755</span></span>
<span><span class="co">## 372 exposedTRUE Clostridium_sp000435835          372  2.298507 0.9054109</span></span>
<span><span class="co">##          lower      upper score_stat pval</span></span>
<span><span class="co">## 134 -4.1773906 -0.1387848         NA   NA</span></span>
<span><span class="co">## 372  0.5239343  4.0730799         NA   NA</span></span></code></pre>
<p>We do not have strong statistical evidence to conclude the
fold-difference in abundance of these two taxa before and after exposure
is any different than the typical fold-difference. Here’s an example
interpretation appropriate for a paper</p>
<p>We estimate that the average abundance of Clostridium_sp900543325 in
metagenomes is 8.7 times greater <em>before</em> commencing dairy work,
when compared to the typical fold-differences in the average abundance
of taxa across these visits (95% CI 1.1–65; $p = $0.21).</p>
<p>In many data analyses, we’d like to run tests for all taxa that we
have measured. We could run robust score tests for every taxon in this
analysis, but it will take a long time to run all tests serially. They
can easily be run in parallel on a computing cluster (for example, if
you are in a VM!).</p>
<p>We show you how this might look in the following chunk. There’s a
full vignette <a href="https://statdivlab.github.io/radEmu/articles/parallel_radEmu.html" class="external-link">here</a>.</p>
<p>DO NOT RUN THIS UNLESS YOU HAVE THE TIME AND COMPUTING RESOURCES!</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ncores</span> <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">emuTest</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">category</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">score_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="op">~</span> <span class="va">exposed</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">meta</span>,</span>
<span>                       fitted_model <span class="op">=</span> <span class="va">hdw_fit</span>,</span>
<span>                       refit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       cluster <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">host_subject_id</span>,</span>
<span>                       test_kj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>k <span class="op">=</span> <span class="va">covariate_to_test</span>, </span>
<span>                                            j <span class="op">=</span> <span class="va">category</span><span class="op">)</span>, </span>
<span>                       Y <span class="op">=</span> <span class="va">coverages</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">!=</span> <span class="st">"windows"</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"GITHUB_ACTIONS"</span><span class="op">)</span>, <span class="st">"true"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># run if we are on a Mac or Linux machine</span></span>
<span>  <span class="va">score_res</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span>,</span>
<span>                      <span class="va">emuTest</span>,</span>
<span>                      mc.cores <span class="op">=</span> <span class="va">ncores</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co"># don't run if we are on a Windows machine, or if testing with GitHub actions</span></span>
<span>  <span class="co"># You can do it! You just would need a different setup. Please reach out!</span></span>
<span>  <span class="va">score_res</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="co">## robust score test p-value for the first taxon</span></span>
<span>  <span class="va">score_res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co">## robust score test p-value for the second taxon</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">full_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">)</span>, </span>
<span>                       <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">score_res</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">score_stat</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">full_pval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">score_res</span><span class="op">)</span>, </span>
<span>                      <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">score_res</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">full_coef</span> <span class="op">&lt;-</span> <span class="va">hdw_fit</span><span class="op">$</span><span class="va">coef</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">score_stat</span>, <span class="op">-</span><span class="va">pval</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">category_num</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">coverages</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>score_stat <span class="op">=</span> <span class="va">full_score</span>,</span>
<span>           pval <span class="op">=</span> <span class="va">full_pval</span><span class="op">)</span></span>
<span>  <span class="va">full_coef</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>By the way – running this many tests is quite reasonable, but the
computation time increases as the number of predictors and number of
taxa increases. When you need it, the package <code>fastEmu</code> will
be ready for you &lt;3 We recommend using <code>fastEmu</code> over
<code>radEmu</code> when running differential abundance analyses for
very large data sets. That said, you do need to choose a reference set
to use <code>fastEmu</code>, which makes the interpretation of parameter
estimates less intuitive than for <code>radEmu</code>. If you find that
<code>radEmu</code> is taking too long for your differential abundance
analysis, consider trying <code>fastEmu</code>!</p>
</div>
<div class="section level2">
<h2 id="a-quick-constrast-with-other-methods">A quick constrast with other methods<a class="anchor" aria-label="anchor" href="#a-quick-constrast-with-other-methods"></a>
</h2>
<p><code>radEmu</code> isn’t the only differential abundance method.
Some other popular alternatives include <code>ALDEx2</code>,
<code>ANCOM-BC2</code>, and <code>DESeq2</code>. Each of these methods
(and the many differential abundance methods that we didn’t consider
here) have pros and cons. Here’s a list of our opinions on these
methods. <strong>These critiques are data-driven, but definitely reflect
our bias.</strong></p>
<ul>
<li>
<code>radEmu</code>
<ul>
<li>pros: low bias, has Type I error rate control in all settings that
we tested, handles sparsity well</li>
<li>cons: robust score tests are slow, especially in data sets with a
large number of categories (taxa, genes, etc.)</li>
</ul>
</li>
<li>
<code>fastEmu</code>
<ul>
<li>pros: inherits <code>radEmu</code> pros, is much faster than
<code>radEmu</code> (especially with a large number of categories)</li>
<li>cons: an approximate method (rather than an exact method) with a
less intuitive target parameter, still slower than several other
methods</li>
</ul>
</li>
<li>
<code>ALDEx2</code>
<ul>
<li>pros: very biased/variable estimates, has Type I error rate control
in many situations, faster than <code>radEmu</code> to test all
categories</li>
<li>cons: has very low power, handles zeroes poorly/weirdly</li>
</ul>
</li>
<li>
<code>ANCOM-BC2</code>
<ul>
<li>pros: low bias, high power, faster than <code>radEmu</code> to test
all categories</li>
<li>cons: terrible error rate control (i.e. not a valid test), can’t
handle data separation (a common consequence of sparsity), relies on
questionable sensitivity analysis</li>
</ul>
</li>
<li>
<code>DESeq2</code>
<ul>
<li>pros: very biased/variable estimates, faster than
<code>radEmu</code> to test all categories</li>
<li>cons: fails to control Type I error in some situations, tailored for
RNA-seq data, not microbiome data, its own developers recommend against
it for microbiome data</li>
</ul>
</li>
<li>
<code>LinDA</code>
<ul>
<li>pros: faster than <code>radEmu</code> to test all categories, high
power in some settings</li>
<li>cons: biased for largest effect sizes, fails to control Type I error
in some situations (large sample sizes with sparse data), handles zeroes
poorly/weirdly</li>
</ul>
</li>
</ul>
<p>Check out the documentation for each of these packages for more
information and more practice using them. Although this example uses a
single binary covariate, each of these methods can be run with more
complex regression models.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amy Willis, Sarah Teichman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
