<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EBAME10: Functional enrichment in anvi'o • blog</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EBAME10: Functional enrichment in anvi'o">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ebame10_regression_pt1.html">EBAME10: Functional enrichment in anvi'o</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt2.html">EBAME10: Functional enrichment in R</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_extn_spaces.html">PNF: Extension spaces for phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_radEmu.html">PNF: radEmu for differential abundance</a></li>
    <li><a class="dropdown-item" href="../articles/stamps_2025.html">STAMPS 2025: How to write a great application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>EBAME10: Functional enrichment in anvi'o</h1>
                        <h4 data-toc-skip class="author">Amy D
Willis</h4>
            
            <h4 data-toc-skip class="date">October 17 2025</h4>
      

      <div class="d-none name"><code>ebame10_regression_pt1.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This lab covers “functional enrichment” in anvi’o, in its simplest
form.</p>
<p>The main learning objective is for you to deepen your understanding
of regression models as tools to answer questions about your
metagenomics data. See the <a href="https://github.com/statdivlab/presentations/blob/main/ebame10/2-regression.pdf" class="external-link">lecture
slides</a>.</p>
<p>Functional enrichment is an example of estimating comparative
parameters in metagenomics, so it gives us the chance to illustrate some
concepts from lecture.</p>
<p>Functional enrichment fits broadly into the paradigm of pangenomics.
If you want to learn about the amazing pangenomics tools in anvi’o,
check out the tutorials at <a href="https://anvi'o.org/learn/" class="external-link">anvi’o.org/learn/</a>.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Navigate to wherever you have been running anvi’o the last few days.
Activate your anvi’o conda environment if needed. Confirm anvi’o
commands run (e.g., <code>anvi-migrate -h</code>). If they don’t, fix
it. These commands work on anvi’o <code>v8</code> (or <code>dev</code>
as of October 2025).</p>
<p>We’re going to work with data from the amazing (now a little
outdated) pangenomics in anvi’o tutorial available <a href="https://merenlab.org/2016/11/08/pangenomics-v2/" class="external-link">here</a>.</p>
<p>Download the zipped data, unzip it, and move into that directory.</p>
<pre><code>wget https://ndownloader.figshare.com/files/28834476 -O Prochlorococcus_31_genomes.tar.gz
tar -zxvf Prochlorococcus_31_genomes.tar.gz
cd Prochlorococcus_31_genomes</code></pre>
<p>Here we have some Prochlorococcus genomes in contigs DBs. How many
genomes are there?</p>
<pre><code><span><span class="va">ls</span> <span class="op">*</span><span class="va">.db</span> <span class="op">|</span> <span class="va">wc</span> <span class="op">-</span><span class="va">l</span> </span></code></pre>
<p>Each was recovered from either a high-light ocean sample, or a
low-light ocean sample. You can read about them in way more detail in
the tutorial linked above.</p>
<p>We want to identify which COGs were present more often in one type of
genomes. As discussed in the lecture, this is equivalent to asking
whether the odds of detecting a COG in a genome differs based on where
the genome was recovered from.</p>
<p>We actually don’t need to build a pangenome to answer this question*.
All we need is</p>
<ol style="list-style-type: decimal">
<li>A list of the genomes that we want to include in the analysis. Here,
we’re going to use the external genomes file
<code>external-genomes.txt</code>. (Check it out its format with
<code>head external-genomes.txt</code>.)</li>
<li>The name of the file we want to save the results. I’m going to call
it <code>enrichment-output.txt</code>.</li>
<li>The annotation source. I’m going to choose
<code>COG14_FUNCTION</code>.</li>
<li>The groups we want to compare across. This info is contained in
<code>layer-additional-data.txt</code>.</li>
</ol>
<p>*You could, of course! It allows you to do lots of interesting
things! It can take a while, though. The approach presented here is
faster.</p>
</div>
<div class="section level2">
<h2 id="attempting-to-run-anvi-compute-functional-enrichment-across-genomes">Attempting to run
<code>anvi-compute-functional-enrichment-across-genomes</code><a class="anchor" aria-label="anchor" href="#attempting-to-run-anvi-compute-functional-enrichment-across-genomes"></a>
</h2>
<p>One of the many cool things about anvi’o is that it’s hard to break
stuff. So, as a general rule, I try things until my wish comes true.
Often, the help messages are enough to get me there. Let’s see if they
are enough today!</p>
<pre><code>anvi-compute-functional-enrichment-across-genomes -e external-genomes.txt -o enrichment-output.txt --annotation-source COG14_FUNCTION </code></pre>
<p>“error: the following arguments are required: -G/–groups-txt”. Ok,
fine. That info is in <code>layer-additional-data.txt</code>:</p>
<pre><code>head layer-additional-data.txt</code></pre>
<p>Ok, maybe this will work?</p>
<pre><code>anvi-compute-functional-enrichment-across-genomes -e external-genomes.txt -o enrichment-output.txt --annotation-source COG14_FUNCTION -G layer-additional-data.txt</code></pre>
<p>“Config Error: A groups-txt file should have a column that is called
<code>group</code>.”</p>
<p>Ok, cool! <code>layer-additional-data.txt</code> has what we need,
but not in a column called group.</p>
<p>I hate remembering stuff, so of course, I asked my bff ChatGPT.</p>
<pre><code>I have a file called layer-additional-data.txt. here's what it looks like 

head layer-additional-data.txt
isolate clade   light
AS9601  HL_II   HL
CCMP1375    LL_II   LL
EQPAC1  HL_I    HL
GP2 HL_II   HL
LG  LL_II   LL
MED4    HL_I    HL
MIT9107 HL_II   HL
MIT9116 HL_II   HL
MIT9123 HL_II   HL

I want to copy the column light and call it group. give me a cmd line one-liner thx</code></pre>
<p>(You don’t have to say thanks, but maybe we’ll be less likely to turn
into paperclips)</p>
<pre><code>awk 'BEGIN{OFS="\t"} NR==1{$(NF+1)="group"} NR&gt;1{$(NF+1)=$3}1' layer-additional-data.txt &gt; layer-additional-data-with-group.txt\n</code></pre>
<p>Ok! Let’s try again!</p>
<pre><code>anvi-compute-functional-enrichment-across-genomes -e external-genomes.txt -o enrichment-output.txt --annotation-source COG14_FUNCTION -G layer-additional-data-with-group.txt</code></pre>
<p>“Config Error: The database at ‘Prochlorococcus_31_genomes/AS9601.db’
is outdated (this database is v20 and your anvi’o installation wants to
work with v24). You can migrate your database without losing any data
using the program <code>anvi-migrate</code> with either of the flags
<code>--migrate-safely</code> or <code>--migrate-quickly</code>.”</p>
<p>Ok, so…</p>
<pre><code><span><span class="va">anvi</span><span class="op">-</span><span class="va">migrate</span> <span class="op">*</span><span class="va">.db</span> <span class="op">-</span><span class="op">-</span><span class="va">migrate</span><span class="op">-</span><span class="va">safely</span></span></code></pre>
<p>And now…</p>
<pre><code>anvi-compute-functional-enrichment-across-genomes -e external-genomes.txt -o enrichment-output.txt --annotation-source COG14_FUNCTION -G layer-additional-data-with-group.txt</code></pre>
<p>Hooray!!! Now let’s open up <code>enrichment-output.txt</code> and
talk about it.</p>
</div>
<div class="section level2">
<h2 id="can-i-just-have-the-answer">Can I just have the answer?<a class="anchor" aria-label="anchor" href="#can-i-just-have-the-answer"></a>
</h2>
<p>Ok :’)</p>
<pre><code>wget https://ndownloader.figshare.com/files/28834476 -O Prochlorococcus_31_genomes.tar.gz
tar -zxvf Prochlorococcus_31_genomes.tar.gz
cd Prochlorococcus_31_genomes
awk 'BEGIN{OFS="\t"} NR==1{$(NF+1)="group"} NR&gt;1{$(NF+1)=$3}1' layer-additional-data.txt &gt; layer-additional-data-with-group.txt\n
anvi-migrate *.db --migrate-safely
anvi-compute-functional-enrichment-across-genomes -e external-genomes.txt -o enrichment-output.txt --annotation-source COG14_FUNCTION -G layer-additional-data-with-group.txt</code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amy Willis, Sarah Teichman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
