<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EBAME10: Functional enrichment in R • blog</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="EBAME10: Functional enrichment in R">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ebame10_regression_pt1.html">EBAME10: Functional enrichment in anvi'o</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt2.html">EBAME10: Functional enrichment in R</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_extn_spaces.html">PNF: Extension spaces for phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_radEmu.html">PNF: radEmu for differential abundance</a></li>
    <li><a class="dropdown-item" href="../articles/stamps_2025.html">STAMPS 2025: How to write a great application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>EBAME10: Functional enrichment in R</h1>
                        <h4 data-toc-skip class="author">Amy D
Willis</h4>
            
            <h4 data-toc-skip class="date">October 17 2025</h4>
      

      <div class="d-none name"><code>ebame10_regression_pt2.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>First, work through the “EBAME10: Functional enrichment in
anvi’o”.</p>
<p>As discussed in lecture, the anvio functional enrichment toolkit is
great for quick and simple comparisons. However, should want more
details! How much do the odds of detecting a gene differ between the
groups? What about when you want to make more complex comparisons? What
about if you studied multiple sites longitudinally?</p>
<p>This tutorial walks you through pulling the data out of anvio, into
R, and running functional enrichment analyses yourself in R.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Open a terminal wherever you were running Part 1. You’re probably in
a directory called <code>Prochlorococcus_31_genomes</code>. I assume
anvio commands can still run.</p>
<p>Remember that we want to ask “How does the odds that a gene is
present differ between high and low light genomes?” To answer this, we
need to know which genes were detected in which genomes. So, let’s get
that info out of anvio and into a format we can load into R.</p>
<p>Let’s make a new directory to keep this info in.</p>
<pre><code>mkdir genes </code></pre>
<p>If you’ve setup your annotation databases correctly, you can pull out
the gene annotations and save them in the folder <code>genes</code> by
copying the following loop into your terminal:</p>
<pre><code>for file in *.db; do
    base=$(basename "$file" .db)
    anvi-export-functions -c "$file" -o "genes/${base}.txt"
done</code></pre>
<p>Ok! Time to open it in R. Open your favourite interactive R editor.
Create a variable with the file path where the genomes DBs are. Load in
the high/low light data.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="st">"presentations/2510-ebame/Prochlorococcus_31_genomes/"</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"layer-additional-data.txt"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="asking-mv-to-clean-up-the-mess-that-is-below-this">Asking MV to clean up the mess that is below this :)<a class="anchor" aria-label="anchor" href="#asking-mv-to-clean-up-the-mess-that-is-below-this"></a>
</h2>
<p>load raoBust. You may need to install it if you haven’t.</p>
<pre><code><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"statdivlab/raoBust"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">raoBust</span><span class="op">)</span></span></code></pre>
<p>Generate presence-absence data</p>
<pre><code><span><span class="va">what_is_there</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"genes"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"*.txt"</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">read_tsv</span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">mutate</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span>, <span class="st">".txt"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">source</span> <span class="op">==</span> <span class="st">"COG14_FUNCTION"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">source</span>, <span class="op">-</span><span class="va">`e_value`</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">rename</span><span class="op">(</span>cog_function <span class="op">=</span> <span class="va">`function`</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">distinct</span></span></code></pre>
<p>Create a look-up table to connect accessions to functions.</p>
<pre><code><span><span class="va">cog_fns</span> <span class="op">&lt;-</span> <span class="va">what_is_there</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">accession</span>, <span class="va">cog_function</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">distinct</span></span>
<span><span class="va">cog_fns</span></span></code></pre>
<p>Now, make the zero rows:</p>
<pre><code><span><span class="va">genes_long</span> <span class="op">&lt;-</span> <span class="va">what_is_there</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>presence <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">sample</span>,<span class="va">accession</span>,<span class="va">presence</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="va">distinct</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>values_from <span class="op">=</span> <span class="va">presence</span>, values_fill <span class="op">=</span> <span class="fl">0</span>, names_from <span class="op">=</span> <span class="va">accession</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">!</span><span class="va">sample</span>,names_to <span class="op">=</span> <span class="st">"cog"</span>, values_to <span class="op">=</span> <span class="st">"presence"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">inner_join</span><span class="op">(</span><span class="va">meta</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span> <span class="op">=</span> <span class="st">"isolate"</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
<p>Run the model once:</p>
<pre><code><span><span class="va">genes_long</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">glm_test</span><span class="op">(</span><span class="va">presence</span> <span class="op">~</span> <span class="va">light</span>, data <span class="op">=</span> <span class="va">.</span>,</span>
<span>           family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Explain the output. Why does it make sense that the p-value is
large?</p>
<pre><code><span><span class="va">genes_long</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cog</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">presence</span>,<span class="va">light</span><span class="op">)</span> </span></code></pre>
<p>Everyone has this gene! And that makes sense, because…</p>
<pre><code><span><span class="va">cog_fns</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">accession</span> <span class="op">==</span> <span class="st">"COG0592"</span><span class="op">)</span> </span></code></pre>
<p>Ok, how do we run this at scale?</p>
<pre><code><span><span class="va">coefs</span> <span class="op">&lt;-</span> <span class="va">genes_long</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_split</span><span class="op">(</span><span class="va">cog</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="va">head</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">set_names</span><span class="op">(</span><span class="fu">map_chr</span><span class="op">(</span><span class="va">.</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">.x</span><span class="op">$</span><span class="va">cog</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">map</span><span class="op">(</span><span class="op">~</span> <span class="fu">glm_test</span><span class="op">(</span><span class="va">presence</span> <span class="op">~</span> <span class="va">light</span>, data <span class="op">=</span> <span class="va">.x</span>, family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"logit"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">imap_dfr</span><span class="op">(</span><span class="op">~</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">.x</span>, rownames <span class="op">=</span> <span class="st">"term"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>             <span class="fu">mutate</span><span class="op">(</span>cog <span class="op">=</span> <span class="va">.y</span><span class="op">)</span>,</span>
<span>           .id <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">relocate</span><span class="op">(</span><span class="va">cog</span><span class="op">)</span></span></code></pre>
<p>Now, look at the results!</p>
<pre><code><span><span class="va">coefs</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">`Non-robust Std Error`</span>, <span class="op">-</span><span class="va">`Non-robust Wald p`</span>, <span class="op">-</span><span class="va">`Robust Wald p`</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Estimate</span>, y <span class="op">=</span> <span class="va">`Robust Std Error`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">out</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">==</span> <span class="st">"lightLL"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amy Willis, Sarah Teichman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
