<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PNF: Extension spaces for phylogenetic trees • blog</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PNF: Extension spaces for phylogenetic trees">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blog</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ebame10_regression_pt1.html">EBAME10: Functional enrichment in anvi'o</a></li>
    <li><a class="dropdown-item" href="../articles/ebame10_regression_pt2.html">EBAME10: Functional enrichment in R</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_extn_spaces.html">PNF: Extension spaces for phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/pnf_radEmu.html">PNF: radEmu for differential abundance</a></li>
    <li><a class="dropdown-item" href="../articles/stamps_2025.html">STAMPS 2025: How to write a great application</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>PNF: Extension spaces for phylogenetic trees</h1>
                        <h4 data-toc-skip class="author">Amy D
Willis</h4>
            
            <h4 data-toc-skip class="date">February 7 2025</h4>
      

      <div class="d-none name"><code>pnf_extn_spaces.Rmd</code></div>
    </div>

    
    
<p><a href="https://ieeexplore.ieee.org/abstract/document/10832388" class="external-link">Published
paper</a> | <a href="https://github.com/statdivlab/ExtnSpaces" class="external-link">Software</a> | <a href="https://github.com/statdivlab/ExtnSpaces_supplementary" class="external-link">Supplementary</a>
| <a href="https://arxiv.org/abs/2407.00227" class="external-link">Preprint</a></p>
<p><strong>TL;DR</strong>: The StatDivLab has a big vision to
revolutionize how you estimate bacterial, archaeal, and interdomain
phylogenetic trees. This paper is our first step in this direction. This
method is quite specific, and may not be feasible for your analysis, but
it reflects some ideas we’re excited about and are advancing further.
Stay tuned.</p>
<p><strong>Long term plan</strong>: Better estimates of bacterial
phylogenies, with accurate and clear uncertainty measures. Address this
“100% bootstrap support for splits supported by no genes” nonsense. If
you don’t believe me, here’s Supplementary Data Figure 7 from <a href="https://academic.oup.com/biostatistics/article/25/3/786/7332152" class="external-link">our
“groves” paper</a>:</p>
<div class="float">
<img src="support.png" alt="Caption: The Prevotella phylogenomic tree with splits colored by bootstrap support (left) and gene tree support (right). We observe zero gene tree support for several deep splits, but note the high bootstrap support for these splits."><div class="figcaption">Caption: The Prevotella phylogenomic tree with
splits colored by bootstrap support (left) and gene tree support
(right). We observe zero gene tree support for several deep splits, but
note the high bootstrap support for these splits.</div>
</div>
<p><strong>This paper</strong>: A distance between phylogenetic trees
with different leaf sets</p>
<p><strong>Why is this important</strong>: Different genes provide
complementary information about shared ancestry. I believe that while we
should be selective about what genes we use to estimate phylogenies, we
shouldn’t be concatenating them together and pretending it’s one
supergene. Instead, we should estimate gene-level phylogenies and
combine them cleverly. But, not all genomes will have all core genes
(e.g., imperfect assemblies, but other reasons too), and we still care
about those genomes. We therefore have a collection of phylogenetic
trees to combine… but they don’t have exactly the same leaves. How do
you compare trees that don’t have the same leaves? I like to compare
trees that <em>do</em> have the same leaves using the BHV distance.
That’s why we came up with an algorithm that extends the BHV distance to
accommodate non-identical leaf sets.</p>
<p><strong>My favourite things</strong>: Resolves (IMO) the biggest
limitation of BHV tree space for analyzing bacterial gene trees. Paves
the way for faster approaches. Creates the first data-analytic tools for
non-identical leaf-ed trees in BHV space.</p>
<p><strong>My least favourite thing</strong>: There are so many dang
tree topologies that it’s not really computationally feasible in
practice. Also, the geometry of extension spaces means the distance is
not formally a metric (both positivity and triangle inequality RIP).</p>
<p><strong>Use cases</strong>: Rather limited. Your typical microbial
ecologist probably won’t use this. They might use <a href="https://academic.oup.com/biostatistics/article/25/3/786/7332152" class="external-link">this</a>,
though. Target audience is more likely BHV space supernerds, convex
optimisation nerds, phylogenetic tree distance afficionados.</p>
<p><strong>Top shoutouts</strong>: We leaned heavily on <a href="https://epubs.siam.org/doi/10.1137/18M1235855" class="external-link">this</a> work from
Megan Owen and Gillian Grindstaff. I still love <a href="https://academic.oup.com/biostatistics/article/25/3/786/7332152" class="external-link">this
paper</a> on tree visualization with Sarah Teichman and Mike Lee.</p>
<p><strong>Review process</strong>: Amazingly smooth. Sparkly mineral
water in a desert wasteland. Reviewers were constructive, reasonable and
fast. I’ll definitely submit to IEEE TCBB again.</p>
<p><strong>Misc</strong>: Dr Maria Valdez was awarded the School of
Public Health Outstanding PhD Award for this work. Go Maria! Thoroughly
well-deserved.</p>
<p><strong>Wrap-up</strong>: Stay tuned! Practical tools for bacterial
phylogenetics are in the works…</p>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amy Willis, Sarah Teichman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
